File: Design-Patterns-Using-JavaScript/Decorator/DecoratorPatternCoffee.js
// Coffee (Component)
class Coffee {
    getCost() {
        throw new Error("Abstract method: getCost");
    }

    getIngredients() {
        throw new Error("Abstract method: getIngredients");
    }
}

// SimpleCoffee (ConcreteComponent)
class SimpleCoffee extends Coffee {
    getCost() {
        return 10;
    }

    getIngredients() {
        return "Coffee";
    }
}

// CoffeeDecorator (Decorator)
class CoffeeDecorator extends Coffee {
    constructor(component) {
        super();
        this.component = component;
    }

    getCost() {
        throw new Error("Abstract method: getCost");
    }

    getIngredients() {
        throw new Error("Abstract method: getIngredients");
    }
}

// MilkDecorator (ConcreteDecorator)
class MilkDecorator extends CoffeeDecorator {
    constructor(component) {
        super(component);
    }

    getCost() {
        return this.component.getCost() + 4;
    }

    getIngredients() {
        return this.component.getIngredients() + ", Milk";
    }
}

// EspressoDecorator (ConcreteDecorator)
class EspressoDecorator extends CoffeeDecorator {
    constructor(component) {
        super(component);
    }

    getCost() {
        return this.component.getCost() + 5;
    }

    getIngredients() {
        return this.component.getIngredients() + ", Espresso";
    }
}

// Client code
const component = new SimpleCoffee();
const decorator1 = new MilkDecorator(component);
const decorator2 = new EspressoDecorator(decorator1);

console.log("Coffee cost is :: " + decorator2.getCost());
console.log("Coffee ingredients are :: " + decorator2.getIngredients());

const latte = new MilkDecorator(new MilkDecorator(new SimpleCoffee()));
console.log("Coffee cost is :: " + latte.getCost());
console.log("Coffee ingredients are :: " + latte.getIngredients());

/*
Coffee cost is :: 19
Coffee ingredients are :: Coffee, Milk, Espresso
Coffee cost is :: 18
Coffee ingredients are :: Coffee, Milk, Milk
*/

File: Design-Patterns-Using-Ruby/Decorator/DecoratorPatternCoffee.rb
 
# Coffee (Component)
class Coffee
    def get_cost
    end
    
    def get_ingredients
    end
end

# SimpleCoffee (ConcreteComponent)
class SimpleCoffee < Coffee
    def get_cost
        10
    end

    def get_ingredients
        'Coffee'
    end
end

# CoffeeDecorator (Decorator)
class CoffeeDecorator < Coffee
    def initialize(component)
        @component = component
    end

    def get_cost
        @component.get_cost
    end

    def get_ingredients
        @component.get_ingredients
    end
end

# MilkDecorator (ConcreteDecorator)
class MilkDecorator < CoffeeDecorator
    def initialize(component)
        super(component)
    end

    def get_cost
        @component.get_cost + 4
    end

    def get_ingredients
        @component.get_ingredients + ', Milk'
    end
end

# EspressoDecorator (ConcreteDecorator)
class EspressoDecorator < CoffeeDecorator
    def initialize(component)
        super(component)
    end

    def get_cost
        @component.get_cost + 5
    end

    def get_ingredients
        @component.get_ingredients + ', Espresso'
    end
end

# Client code
component = SimpleCoffee.new
decorator1 = MilkDecorator.new(component)
decorator2 = EspressoDecorator.new(decorator1)

puts "Coffee cost is :: #{decorator2.get_cost}"
puts "Coffee ingredients are :: #{decorator2.get_ingredients}"

latte = MilkDecorator.new(MilkDecorator.new(SimpleCoffee.new))
puts "Coffee cost is :: #{latte.get_cost}"
puts "Coffee ingredients are :: #{latte.get_ingredients}"

=begin 
Coffee cost is :: 19
Coffee ingredients are :: Coffee, Milk, Espresso
Coffee cost is :: 18
Coffee ingredients are :: Coffee, Milk, Milk
=end
